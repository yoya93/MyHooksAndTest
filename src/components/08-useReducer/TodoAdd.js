import React from "react";
import { useForm } from "../../hooks/useForm";

export const TodoAdd = ({ handleAdd }) => {
  const [value, handleInputChange, reset] = useForm({
    description: "",
  });

  const { description } = value;

  const handleSubmitAdd = (e) => {
    e.preventDefault();

    if (description.trim().length < 1) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false,
    };

    handleAdd(newTodo);

    reset();
  };

  return (
    <>
      <h3>Agregar Todo</h3>
      <hr />

      <form onSubmit={handleSubmitAdd}>
        <input
          className="form-control"
          value={description}
          name="description"
          type="text"
          placeholder="Aprender ....."
          autoComplete="off"
          onChange={handleInputChange}
        ></input>

        <button
          type="submit"
          className="btn btn-outline-primary btn-block mt-1"
        >
          Agregar
        </button>
      </form>
    </>
  );
};
