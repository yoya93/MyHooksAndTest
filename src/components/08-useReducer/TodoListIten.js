import React from "react";

export const TodoListIten = ({ i, todo, handleThrough, handleDeleteIten }) => {
  return (
    <>
      <li key={todo.id} className="list-group-iten mb-2">
        <p
          onClick={() => {
            handleThrough(todo.id);
          }}
          className={`${todo.done && "complete"}`}
        >
          {i + 1}-{todo.desc}
        </p>
        <button
          onClick={() => {
            handleDeleteIten(todo.id);
          }}
          className="btn btn-danger"
        >
          Borrar
        </button>
      </li>
    </>
  );
};
