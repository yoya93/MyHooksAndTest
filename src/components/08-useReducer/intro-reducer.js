const initialState = [
  {
    id: 1,
    todo: "Comprar leche",
    done: false,
  },
];

const todoReducer = (state = initialState, AddTodoAction) => {
  if (AddTodoAction?.type === "agregar") {
    return [...state, AddTodoAction.pyload];
  }
  return state;
};

let todos = todoReducer();

const newTod = {
  id: 2,
  todo: "Comprar Pan",
  done: false,
};

const AddTodoAction = {
  type: "agregar",
  pyload: newTod,
};

todos = todoReducer(todos, AddTodoAction);

console.log(todos);
