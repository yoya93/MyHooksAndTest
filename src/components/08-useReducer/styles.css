body {
  padding: 50px;
}

p {
  cursor: pointer;
  margin: 0px;
}

.list-group-iten {
  align-items: center;
  display: flex;
  justify-content: space-between;
}

.complete {
  text-decoration: line-through;
}
