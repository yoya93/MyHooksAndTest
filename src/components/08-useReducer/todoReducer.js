export const todoReducer = (state = [], action) => {
  switch (action.type) {
    case "add":
      return [...state, action.payload];
    case "delete":
      return state.filter((todo) => todo.id !== action.payload); //filter regresa un arreglo con todos los itens que cumplan esa condicion
    case "through":
      return state.map((todo) =>
        todo.id === action.payload ? { ...todo, done: !todo.done } : todo
      );

    default:
      return state;
  }
};

// case "through":
//   return state.map((todo) => {
//     if (todo.id === action.payload) {
//       // busco el elemento del array ke se kiere cambiar
//       return { ...todo, done: !todo.done }; //retorno el objeto con la propiedad done cambiada
//     } else {
//       return todo;
//     }
//   });
