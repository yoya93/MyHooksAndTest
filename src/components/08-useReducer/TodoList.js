import React from "react";
import { TodoListIten } from "./TodoListIten";

export const TodoList = ({ todos, handleThrough, handleDeleteIten }) => {
  return (
    <>
      <ul className="list-group list-group-flush">
        {todos.map((todo, i) => (
          <TodoListIten
            key={todo.id}
            todo={todo}
            i={i}
            handleDeleteIten={handleDeleteIten}
            handleThrough={handleThrough}
          />
        ))}
      </ul>
    </>
  );
};
