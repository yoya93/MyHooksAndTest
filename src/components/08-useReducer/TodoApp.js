import React, { useReducer, useEffect } from "react";
import { todoReducer } from "./todoReducer";

import { TodoAdd } from "./TodoAdd";
import { TodoList } from "./TodoList";

import "./styles.css";

const init = () => {
  return JSON.parse(localStorage.getItem("todos")) || [];
  //[
  //     {
  //       id: new Date().getTime(),
  //       desc: "Aprender React",
  //       done: false,
  //     },
  //   ];
};

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  const handleDeleteIten = (id) => {
    const action = {
      type: "delete",
      payload: id,
    };

    dispatch(action);
  };

  const handleThrough = (id) => {
    dispatch({
      type: "through",
      payload: id,
    });
  };

  const handleAdd = (newTodo) => {
    dispatch({
      type: "add",
      payload: newTodo,
    });
  };

  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos));
  }, [todos]);

  return (
    <>
      <h1>TodoApp({todos.length})</h1>
      <hr />

      <div className="row">
        <div className="col-7">
          <TodoList
            handleDeleteIten={handleDeleteIten}
            handleThrough={handleThrough}
            todos={todos}
          />
        </div>

        <div className="col-5">
          <TodoAdd handleAdd={handleAdd} />
        </div>
      </div>
    </>
  );
};
