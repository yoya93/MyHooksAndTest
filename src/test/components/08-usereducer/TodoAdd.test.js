import React from "react";
import { TodoAdd } from "../../../components/08-useReducer/TodoAdd";
import { todosDemo } from "../../fixtures";
import "@testing-library/jest-dom";
import { shallow } from "enzyme";

describe("Pruebas en <TodoAdd /> ", () => {
  const handleAdd = jest.fn(); // monitorizacion de mi funcion en el componente
  const wrapper = shallow(<TodoAdd handleAdd={handleAdd} />);

  test("debe de mostrarse correctamente ", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test(" No debe de llamar a  handleAdd ", () => {
    const formSubmit = wrapper.find("form").prop("onSubmit");

    formSubmit({ preventDefault() {} });

    expect(handleAdd).toHaveBeenCalledTimes(0);
  });

  test("Debe de llamar a funcion handleAdd ", () => {
    const value = "Test 1";
    wrapper.find("input").simulate("change", {
      target: {
        value,
        name: "description",
      },
    });

    const formSubmit = wrapper.find("form").prop("onSubmit");

    formSubmit({ preventDefault() {} });

    expect(handleAdd).toHaveBeenCalledTimes(1);
    expect(handleAdd).toHaveBeenCalledWith({
      id: expect.any(Number), // el id canbia por lo cual soo vverifico los otros valores y ke este id sea un number
      desc: value,
      done: false,
    });

    expect(wrapper.find("input").prop("value")).toBe(""); //espero que se halla disparado la function reset
  });
});
