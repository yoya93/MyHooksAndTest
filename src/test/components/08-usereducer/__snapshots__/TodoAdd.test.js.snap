// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoAdd />  debe de mostrarse correctamente  1`] = `
<Fragment>
  <h3>
    Agregar Todo
  </h3>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Aprender ....."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary btn-block mt-1"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
